src/libre/print/tree.c

```diff
@@ -40,48 +40,98 @@ fprintf_count(FILE *f, unsigned count)
 	}
 }
 
+static void
+fprintf_flags(FILE *f, enum re_ast_flags flags)
+{
+	fprintf(f, "(f: ");
+	#define PR_FLAG(X, S)						\
+		if (0 != (flags & (RE_AST_FLAG_ ## X))) {		\
+			fprintf(f, "%s", S);				\
+		}
+	
+	PR_FLAG(FIRST_STATE, "F");
+	PR_FLAG(LAST_STATE, "L");
+	PR_FLAG(UNSATISFIABLE, "X");
+	PR_FLAG(NULLABLE, "0");
+
+	#undef PR_FLAG
+		fprintf(f, ")");
+}
+
 static void
 pp_iter(FILE *f, const struct fsm_options *opt, size_t indent, struct ast_expr *n)
 {
 	assert(f != NULL);
 	assert(opt != NULL);
 
 	if (n == NULL) { return; }
-	/* assert(n != NULL); */
+
 	INDENT(f, indent);
 
 	#define IND 4
 
+	fprintf(f, "%p", (void *)n);
+	if (n->flags != 0x00) {
+		fprintf(f, " ");
+		fprintf_flags(f, n->flags);
+	}		
+	fprintf(f, " -- ");
+
 	switch (n->t) {
 	case AST_EXPR_EMPTY:
-		fprintf(f, "EMPTY %p\n", (void *)n);
+		fprintf(f, "EMPTY \n");
 		break;
+
+	/* CONCAT and ALT should be removed once CONCAT_N and ALT_N are
+	 * built directly in the parser. */
 	case AST_EXPR_CONCAT:
 		fprintf(f, "CONCAT %p: {\n", (void *)n);
 		pp_iter(f, opt, indent + 1*IND, n->u.concat.l);
 		INDENT(f, indent);
+
 		fprintf(f, ", (%p)\n", (void *)n);
+
 		pp_iter(f, opt, indent + 1*IND, n->u.concat.r);
 		INDENT(f, indent);
 		fprintf(f, "} (%p)\n", (void *)n);
```

Comment:
Missing `break;` here I think (and below)

Reply:
Good catch. Fixed in 6b56c40.

I deleted these, then added them back later (when I was debugging flattening the linked lists into arrays), but wound up not needing to check them, so I guess I didn't notice them missing the break.

Once the parser builds arrays directly, these cases should just be removed.